### ISSUE TITLE
Errors with . and .data pronouns in do()

### ISSUE THREAD

**Issue by Hong-Revo** [tidyverse/dplyr#2998]
This is with a fresh build of dplyr as of 20 minutes ago. It looks like multiple references to `.` or `.data` is confusing the evaluator.

```
# this works
> do(mtcars, m=lm(mpg ~ ., data=.))
# A tibble: 1 x 1
         m
    <list>
1 <S3: lm>


# these don't work
> do(mtcars, m=lm(mpg ~ ., data=.), w=lm(wt ~ ., data=.))
Error in is.data.frame(data) : object '.' not found

> do(mtcars, m=lm(mpg ~ ., data=.data), w=lm(wt ~ ., data=.data))
Error in is.data.frame(data) : object '.data' not found
```

**Comment by krlmlr** [comment]
I suspect this is due to the forceful cleaning of the data mask in `eval_tidy_()`, which is called by `do()` in a loop. `eval_tidy_()` is now deprecated. What's the best way to proceed? Is #3318 related?

**Comment by lionel-** [comment]
Yes with previous rlang it should have been `overscope_eval_next()` rather than `eval_tidy_()` as the former does not clean the mask.

With the new simplified API you just create an overscope and supply it to `eval_tidy()`. Note that there is an issue with rlang 0.2.0 that will be fixed in 0.2.1, the mask is still cleaned up at each call. So until the next version is out I would use the soft-deprecated `overscope_eval_next()`.

**Comment by krlmlr** [comment]
Thanks, Lionel, your suggestion seems to work. Simpler reprex:

``` r
dplyr::do(mtcars, a = nrow(.), b = nrow(.))
#> Error in nrow(.): object '.' not found
```

> Created on 2018-03-19 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).
